Add data exception codes. Part 1. (#186)
neo4j/docs-status-codes: modules/ROOT/pages/errors/gql-errors.adoc

[cols="<1s,<4"]
|===
|GQLSTATUS code
m|22N12
|Status description
a|error: data exception - invalid date, time, or datetime format. Invalid argument: cannot process `$input`.
|Classification
m|CLIENT_ERROR
|===

=== 22N13 

Notifies the client that a date component was not provided for the specified time zones.

.Error details
[cols="<1s,<4"]
|===
|GQLSTATUS code
m|22N13
|Status description
a|error: data exception - invalid time zone. Specified time zones must include a date component.
|Classification
m|CLIENT_ERROR
|===

=== 22N14 

Notifies the client that the provided data type and its value cannot be combined.

.Error details
[cols="<1s,<4"]
|===
|GQLSTATUS code
m|22N14
|Status description
a|error: data exception - invalid temporal value combination. Cannot select both `$temporal` and `$component`.
|Classification
m|CLIENT_ERROR
|===

=== 22N15 

Notifies the client that the provided data types contain an unsupported value.

.Error details
[cols="<1s,<4"]
|===
|GQLSTATUS code
m|22N15
|Status description
a|error: data exception - invalid temporal component. Cannot read the specified `$component` component from `$temporal`.
|Classification
m|CLIENT_ERROR
|===

=== 22N16 

> Notifies the client that .

Author: ```suggestion
Notifies the client that the wrong Cypher clause is used during data import.
```

.Error details
[cols="<1s,<4"]
|===
|GQLSTATUS code
m|22N16
|Status description
a|error: data exception - invalid import value. Importing entity values to a graph with a `USE` clause is not supported. Attempted to import `$expr` to `$graph`.
|Classification
m|CLIENT_ERROR
|===

=== 22N17 

> Notifies the client that .

Author: ```suggestion
Notifies the client that the provided argument in a temporal function is incorrect.
```

Author: Is this code about instant temporal types?

[[neo4j-gqlstatus-errors]]
= List of all GQLSTATUS server error codes

> The following page provides an overview of all server errors in Neo4j, along with some scenarios and their possible solutions.

Author: ```suggestion
//The following page provides an overview of all server errors in Neo4j, along with some scenarios and their possible solutions.
The following page provides an overview of all GQLSTATUS server error codes in Neo4j.
All errors in Neo4j have severity level `ERROR`.
```

All errors in Neo4j have severity level `ERROR`.

* <<_connection-exceptions, Connection exceptions>>
* <<_data-exceptions, Data exceptions>>

> //The following page provides an overview of all server errors in Neo4j, along with some scenarios and their possible solutions.

Author: ```suggestion
```

> The following page provides an overview of all GQLSTATUS server error codes in Neo4j.

Author: ```suggestion
```

> All errors in Neo4j have severity level `ERROR`.

Author: ```suggestion
```

[[neo4j-gqlstatus-errors]]
= List of all GQLSTATUS server error codes

The following page provides an overview of all server errors in Neo4j, along with some scenarios and their possible solutions.
All errors in Neo4j have severity level `ERROR`.

* <<_connection-exceptions, Connection exceptions>>
* <<_data-exceptions, Data exceptions>>

//The following page provides an overview of all server errors in Neo4j, along with some scenarios and their possible solutions.
The following page provides an overview of all GQLSTATUS server error codes in Neo4j.
All errors in Neo4j have severity level `ERROR`.

[[_connection-exceptions]]
== Connection exceptions

> The following is a list of the connection exceptions that can occur in Neo4j.

Author: ```suggestion
```

=== 08N15

Status description:: error: connection exception - no such routing policy. Policy definition of the routing policy { $routingPolicy } could not be found. Verify that the spelling is correct.

[[_data-exceptions]]
== Data exceptions

> The following is a list of the data exceptions that can occur in Neo4j.

Author: @Lojjs, could you provide an intro sentence for data exceptions?

Reviewer: I can try :) something like this maybe:

> Database exceptions occur when a client request contains wrong format or types, or other unsupported input. Some examples are data and constraint creation which conflicts with already existing constraints, properties of non-storable type, and spatial and temporal values with invalid components. 

This class seems quite broad, but I thought mentioning a few examples might give the reader an idea about which sort of errors are covered.

=== 08N15

Status description:: error: connection exception - no such routing policy. Policy definition of the routing policy { $routingPolicy } could not be found. Verify that the spelling is correct.

[[_data-exceptions]]
== Data exceptions

The following is a list of the data exceptions that can occur in Neo4j.

=== 22N00 

Status description:: error: data exception - unsupported value. The provided value is unsupported and cannot be processed.

=== 22N01 

Status description:: error: data exception - invalid type.

> Expected the value { $value } to be of type { $valueTypeOredList }, but was of type { $valueType }.

Reviewer: The 'Ored' or 'Anded' in some list parameters is mostly a way to show which format it should be written out on. Made up example to show the principle:

The element(s) { $valueTypeList } are broken => "The element(s) a, b**,** c are broken"
The element(s) { $valueTypeOredList } are broken => "The element(s) a, b **or** c are broken"
The element(s) { $valueTypeAndedList } are broken => "The element(s) a, b **and** c are broken"

In the codebase we just call the key "valueTypeList" in all cases and have a separate internal joinstyle thing which tracks which of these three formats should be used. I'm not sure whether it is helpful or just confusing for the user to call it  `{ $valueTypeOredList }` or if we should simply call it `{ $valueTypeList }` like in the code? 

When we add examples later those will show on which format we print out the list in. When we reintroduce the statusParameter in the diagnostic record in a later release (it was descoped for now) that will be a map with just the key and the list as a Java list without any formatting, so there we probably want to call it valueTypeList.
 
 Any thoughts in this @NataliaIvakina and @renetapopova as it applies to a bunch of other places as well.

Author: IMO, `{ $valueTypeList }` is less confusing

Reviewer 2: I agree with Natalia. I have already created a card to our wall to add the parameter descriptions, which I think will eliminate any confusion. How did you come up with 'Ored' and 'Anded' instead of just { $valueTypeOrList } and { $valueTypeAndList }?

Reviewer: Ok let's go for { $valueTypeList } then. This also applies to a few other codes in this PR with list parameters, I think it was like 3 or so more occurrences . 

That is a good question, I think that was Stefan's idea when he did re-designed/formalized how parameter framework works. I guess he is used to writing GQL standard because that whole writing is a bit over-complicated... And then Isak probably used his wording when implementing it.

=== 22N10 

Status description:: error: data exception - invalid pre-parser option value. Invalid pre-parser option, specified { $input } is not valid for option { $option }. Valid options are: { $optionAndedList }.

=== 22N11 

Status description:: error: data exception - invalid argument. Invalid argument: cannot process { $input }.

=== 22N12 

Status description:: error: data exception - invalid date, time, or datetime format. Invalid argument: cannot process { $input }.

=== 22N13 

Status description:: error: data exception - invalid time zone. Specified time zones must include a date component.

=== 22N14 

Status description:: error: data exception - invalid temporal value combination. Cannot select both { $temporal } and { $component }.

=== 22N15 

Status description:: error: data exception - invalid temporal component. Cannot read the specified { $component } component from { $temporal }.

=== 22N16 

Status description:: error: data exception - invalid import value. Importing entity values to a graph with a USE clause is not supported. Attempted to import { $expr } to { $graph }.

> === 22N17 

Reviewer: We have deleted 22N17 from the code and it also is marked as 'DELETE' in the spreadsheet. I think it was a copy of another code or wrong in some other way. Maybe you can remind me @anderson4j ? Anyway, we should probably exclude this one from docs too.

Status description:: error: data exception - invalid date, time, or datetime function field name. Cannot read the specified { $component } component from { $temporal }.

=== 22N18 

Status description:: error: data exception - incomplete spatial value. A { $crs } POINT must contain { $mapKeyAndedList }.

=== 22N19 

Status description:: error: data exception - invalid spatial value. A POINT must contain either 'x' and 'y', or 'latitude' and 'longitude'.

=== 22N20 

> Status description:: error: data exception - invalid spatial value dimensions. Cannot create POINT with { $crs } coordinate reference system (CRS) and { $coordinates } coordinates. Use the equivalent { $crs2 } coordinate reference system instead.

Reviewer: ```suggestion
Status description:: error: data exception - invalid spatial value dimensions. Cannot create POINT with { $dim1 }D coordinate reference system (CRS) and { $value } coordinates. Use the equivalent { $dim2 }D coordinate reference system instead.
```

We noticed that the parameters were bad here, so we have updated it in the code. But seems like I missed to update the spreadsheet, have done that now.

=== 22N21 

Status description:: error: data exception - unsupported coordinate reference system. Unsupported coordinate reference system (CRS): { $crs }.

=== 22N22 

Status description:: error: data exception - invalid spatial value combination. Cannot specify both coordinate reference system (CRS) and spatial reference identifier (SRID).

=== 22N23 

Status description:: error: data exception - invalid latitude value. Cannot create WGS84 POINT with invalid coordinate: { $coordinates }. The valid range for the latitude coordinate is [-90, 90].

=== 22N24 

Status description:: error: data exception - invalid coordinate arguments. Cannot construct a { $valueType } from { $coordinates }.

=== 22N25 

Status description:: error: data exception - invalid temporal arguments. Cannot construct a { $valueType } from { $temporal }.

=== 22N26 

Status description:: error: data exception - unsupported rounding mode. Unknown rounding mode. Valid values are: CEILING, FLOOR, UP, DOWN, HALF_EVEN, HALF_UP, HALF_DOWN, UNNECESSARY.

=== 22N27 

> Status description:: error: data exception - invalid entity type. Invalid input { $input } for { $var }. Expected to be one of { $valueTypeOredList }.

Reviewer: ```suggestion
Status description:: error: data exception - invalid entity type. Invalid input { $input } for { $variable }. Expected to be one of { $valueTypeOredList }.
```

Again, something which we seem to have changed in the code but not the spreadsheet. 'var' is a reserved word in Java/Scala so it was very impractical to use as a variable name, therefore we changed it to 'variable'

=== 08N15

Status description:: error: connection exception - no such routing policy. Policy definition of the routing policy { $routingPolicy } could not be found. Verify that the spelling is correct.

[[_data-exceptions]]
== Data exceptions

Database exceptions occur when a client request contains the wrong format, types, or other unsupported input.
Some examples are data and constraint creation, which conflicts with existing constraints, properties of non-storable type, and spatial and temporal values with invalid components.

=== 22N00 

Status description:: error: data exception - unsupported value. The provided value is unsupported and cannot be processed.

=== 22N01 

Status description:: error: data exception - invalid type.

> Expected the value { $value } to be of type { $valueTypeOredList }, but was of type { $valueType }.

Author: ```suggestion
Expected the value { $value } to be of type { $valueTypeList }, but was of type { $valueType }.
```

=== 22N02 

Status description:: error: data exception - specified negative numeric value. Expected { $option } to be a positive number but found { $value } instead.

=== 22N03 

Status description:: error: data exception - specified numeric value out of range. Expected { $component } to be of type { $valueType } and in the range { $lower }  to { $upper } but found { $value }.

=== 22N04 

> Status description:: error: data exception - invalid input value. Invalid input { $input } for { $context }. Expected one of { $inputOredList }.

Author: ```suggestion
Status description:: error: data exception - invalid input value. Invalid input { $input } for { $context }. Expected one of { $inputList }.
```

=== 22N05

Status description:: error: data exception - input failed validation. Invalid input { $input } for { $context }.

=== 22N06 

Status description:: error: data exception - empty input string. Invalid input. { $option } needs to be specified.

=== 22N07 

Status description:: error: data exception - invalid pre-parser option key. Invalid pre-parser option(s): { $optionList }.

=== 22N08 

Status description:: error: data exception - invalid pre-parser combination. Invalid pre-parser option, cannot combine { $option1 } with { $option2 }.

=== 22N09 

Status description:: error: data exception - conflicting pre-parser combination. Invalid pre-parser option, cannot specify multiple conflicting values for { $option }.

=== 22N10 

> Status description:: error: data exception - invalid pre-parser option value. Invalid pre-parser option, specified { $input } is not valid for option { $option }. Valid options are: { $optionAndedList }.

Author: ```suggestion
Status description:: error: data exception - invalid pre-parser option value. Invalid pre-parser option, specified { $input } is not valid for option { $option }. Valid options are: { $optionList }.
```

=== 22N11 

Status description:: error: data exception - invalid argument. Invalid argument: cannot process { $input }.

=== 22N12 

Status description:: error: data exception - invalid date, time, or datetime format. Invalid argument: cannot process { $input }.

=== 22N13 

Status description:: error: data exception - invalid time zone. Specified time zones must include a date component.

=== 22N14 

Status description:: error: data exception - invalid temporal value combination. Cannot select both { $temporal } and { $component }.

=== 22N15 

Status description:: error: data exception - invalid temporal component. Cannot read the specified { $component } component from { $temporal }.

=== 22N16 

Status description:: error: data exception - invalid import value. Importing entity values to a graph with a USE clause is not supported. Attempted to import { $expr } to { $graph }.

// === 22N17 

// Status description:: error: data exception - invalid date, time, or datetime function field name. Cannot read the specified { $component } component from { $temporal }.

=== 22N18 

> Status description:: error: data exception - incomplete spatial value. A { $crs } POINT must contain { $mapKeyAndedList }.

Reviewer: ```suggestion
Status description:: error: data exception - incomplete spatial value. A { $crs } POINT must contain { $mapKeyList }.
```

=== 22N21 

Status description:: error: data exception - unsupported coordinate reference system. Unsupported coordinate reference system (CRS): { $crs }.

=== 22N22 

Status description:: error: data exception - invalid spatial value combination. Cannot specify both coordinate reference system (CRS) and spatial reference identifier (SRID).

=== 22N23 

Status description:: error: data exception - invalid latitude value. Cannot create WGS84 POINT with invalid coordinate: { $coordinates }. The valid range for the latitude coordinate is [-90, 90].

=== 22N24 

Status description:: error: data exception - invalid coordinate arguments. Cannot construct a { $valueType } from { $coordinates }.

=== 22N25 

Status description:: error: data exception - invalid temporal arguments. Cannot construct a { $valueType } from { $temporal }.

=== 22N26 

Status description:: error: data exception - unsupported rounding mode. Unknown rounding mode. Valid values are: CEILING, FLOOR, UP, DOWN, HALF_EVEN, HALF_UP, HALF_DOWN, UNNECESSARY.

=== 22N27 

> Status description:: error: data exception - invalid entity type. Invalid input { $input } for { $variable }. Expected to be one of { $valueTypeOredList }.

Author: ```suggestion
Status description:: error: data exception - invalid entity type. Invalid input { $input } for { $variable }. Expected to be one of { $valueTypeList }.
```

.Error details
[cols="<1s,<4"]
|===
|GQLSTATUS code
m|22N00
|Status description
a|error: data exception - unsupported value. The provided value is unsupported and cannot be processed.
|Classification
m|CLIENT_ERROR
|===

=== 22N01 

Notifies the client that an unsupported value has been provided in a query.

.Error details
[cols="<1s,<4"]
|===
|GQLSTATUS code
m|22N00
|Status description
a|error: data exception - unsupported value. The provided value is unsupported and cannot be processed.
|Classification
m|CLIENT_ERROR
|===

=== 22N01 

Notifies the client that an unsupported value has been provided in a query.

.Error details
[cols="<1s,<4"]
|===
|GQLSTATUS code
m|22N00
|Status description
a|error: data exception - unsupported value. The provided value is unsupported and cannot be processed.
|Classification
m|CLIENT_ERROR
|===

=== 22N01 

Notifies the client that an unsupported value has been provided in a query.

.Error details
[cols="<1s,<4"]
|===
|GQLSTATUS code
m|22N00
|Status description
a|error: data exception - unsupported value. The provided value is unsupported and cannot be processed.
|Classification
m|CLIENT_ERROR
|===
////

> =======

Author: ```suggestion
```
